src/FsAutoComplete/CodeFixes/RemoveUnnecessaryParentheses.fs

```diff
@@ -0,0 +1,168 @@
+      let! txt = sourceText.GetText range
+
+      let firstChar = txt[0]
+      let lastChar = txt[txt.Length - 1]
+
+      match firstChar, lastChar with
+      | '(', ')' ->
+        let adjusted =
+          match sourceText with
+          | TrailingOpen range -> txt[1 .. txt.Length - 2].TrimEnd()
+
+          | Trim range trim & OffsidesDiff range spaces ->
+            match spaces with
+            | NoShift -> trim txt[1 .. txt.Length - 2]
+            | ShiftLeft spaces -> trim (txt[1 .. txt.Length - 2].Replace("\n" + String(' ', spaces), "\n"))
+            | ShiftRight spaces -> trim (txt[1 .. txt.Length - 2].Replace("\n", "\n" + String(' ', spaces)))
+
+          | _ -> txt[1 .. txt.Length - 2].Trim()
+
+        let newText =
+          let (|ShouldPutSpaceBefore|_|) (s: string) =
+            // ……(……)
+            // ↑↑ ↑
+            (sourceText.TryGetChar(range.Start.IncColumn -1), sourceText.TryGetChar range.Start)
+            ||> Option.map2 (fun twoBefore oneBefore ->
+              match twoBefore, oneBefore, s[0] with
+              | _, _, ('\n' | '\r') -> None
+              | '[', '|', (Punctuation | LetterOrDigit) -> None
+              | _, '[', '<' -> Some ShouldPutSpaceBefore
+              | _, ('(' | '[' | '{'), _ -> None
+              | _, '>', _ -> Some ShouldPutSpaceBefore
+              | ' ', '=', _ -> Some ShouldPutSpaceBefore
+              | _, '=', ('(' | '[' | '{') -> None
+              | _, '=', (Punctuation | Symbol) -> Some ShouldPutSpaceBefore
+              | _, LetterOrDigit, '(' -> None
+              | _, (LetterOrDigit | '`'), _ -> Some ShouldPutSpaceBefore
+              | _, (Punctuation | Symbol), (Punctuation | Symbol) -> Some ShouldPutSpaceBefore
+              | _ -> None)
+            |> Option.flatten
+
+          let (|ShouldPutSpaceAfter|_|) (s: string) =
+            // (……)…
+            //   ↑ ↑
+            sourceText.TryGetChar(range.End.IncColumn 1)
+            |> Option.bind (fun endChar ->
+              match s[s.Length - 1], endChar with
+              | '>', ('|' | ']') -> Some ShouldPutSpaceAfter
+              | _, (')' | ']' | '[' | '}' | '.' | ';' | ',' | '|') -> None
+              | (Punctuation | Symbol), (Punctuation | Symbol | LetterOrDigit) -> Some ShouldPutSpaceAfter
+              | LetterOrDigit, LetterOrDigit -> Some ShouldPutSpaceAfter
+              | _ -> None)
+
+          match adjusted with
+          | ShouldPutSpaceBefore & ShouldPutSpaceAfter -> " " + adjusted + " "
+          | ShouldPutSpaceBefore -> " " + adjusted
+          | ShouldPutSpaceAfter -> adjusted + " "
+          | adjusted -> adjusted
+
+        return
+          [ { Edits = [| { Range = d.Range; NewText = newText } |]
+              File = codeActionParams.TextDocument
+              Title = title
+              SourceDiagnostic = Some d
+              Kind = FixKind.Fix } ]
+
+      | notParens ->
+        System.Diagnostics.Debug.Fail $"%A{notParens} <> ('(', ')')"
```

Comment:
can you log something here instead of failing hard?

Reply:
Hmm, the only way it would really be hit at this point would be if this fix were triggered for the wrong diagnostic ID, so maybe I'll just take it out.